### PR TITLE
Rework the way to communicate backpressure (AKA suspended ingestion)

### LINKED ISSUE #1319: While ingestion is suspended, neither the status page is accessible, nor is a clean shutdown possible.
We just found one of our servers having problems to keep up with persistence up to the point that ingestion got suspended. The exact circumstances are still not clear and @beorn7 is investigating.

Unexpectedly (to me at least), the status page would not load during the whole time, while the query interface was still available. It seems the target manager got into an undefined state as well.

### PR DESCRIPTION
**Pull request by beorn7**
This gives up on the idea to communicate throuh the Append() call (by
either not returning as it is now or returning an error as
suggested/explored elsewhere). Here I have added a Throttled() call,
which has the advantage that it can be called before a whole _batch_
of Append()'s. Scrapes will happen completely or not at all. Same for
rule group evaluations. That's a highly desired behavior (as discussed
elsewhere). The code is even simpler now as the whole ingestion buffer
could be removed.

Logging of throttled mode has been streamlined and will create at most
one message per minute.

Fixes #1319 .

@fabxc This is a sneak preview. Still needs to be benchmarked.

### CODE REVIEW

**Comment by beorn7** [comment]
Benchmarks look good. Feel free to review thoroughly.

**Comment by beorn7** [comment]
525k samples/s still worked after some struggle. I guess it's pretty close to breaking then.

File: retrieval/targetmanager.go [hunk 1]
```diff
@@ -165,6 +165,7 @@ func (tm *TargetManager) Run() {
 	})
 
 	tm.running = true
+	log.Info("Target manager started.")
```
Comment (fabxc) [inline retrieval/targetmanager.go]: There's nothing blocking in the startup of the target manager. Let's just print this line than and remove the one above.

File: storage/local/interface.go [hunk 2]
```diff
@@ -34,6 +34,9 @@ type Storage interface {
 	// from the provided Sample as those labels are considered equivalent to
 	// a label not present at all.
 	Append(*model.Sample)
+	// Throttled returns true if the Storage has too many chunks in memory
+	// already or has too many chunks waiting for persistence.
+	Throttled() bool
```
Comment (fabxc) [inline storage/local/interface.go]: I'm still for `Throttle` as in a question. This is a simply a "should I do X?"-request.
We had alternatives where the implication was that the storage requested anything, which I find misleading.

Or invert the bool and make it `HasCapacity()`?

File: storage/storage.go [hunk 3]
```diff
@@ -35,3 +52,14 @@ func (f Fanout) Append(s *model.Sample) {
 		a.Append(s)
 	}
 }
+
+// Throttled returns true if at least one of the SampleAppenders in the Fanout
+// slice is throttled.
+func (f Fanout) Throttled() bool {
```
Comment (fabxc) [inline storage/storage.go]: I don't think that's the correct behavior. We are using the fanout to write to multiple storages simultaneously. We should not stop the world because one of potentially many is too busy.

In the current state it might work, because the remote storage always returns false and is slower anyway.
But this method should return true iff all sub-appenders are throttled.

On `Append()` it should then only call `Append()` on those sub-appenders that are not busy.

File: storage/local/storage.go [hunk 4]
```diff
@@ -616,6 +609,54 @@ func (s *memorySeriesStorage) Append(sample *model.Sample) {
...
+	for {
+		select {
+		case <-s.throttled:
+			if !timer.Reset(time.Minute) {
+				log.
+					With("chunksToPersist", s.getNumChunksToPersist()).
+					With("maxChunksToPersist", s.maxChunksToPersist).
+					With("memoryChunks", atomic.LoadInt64(&numMemChunks)).
+					With("maxToleratedMemChunks", int(float64(s.maxMemoryChunks)*toleranceFactorMemChunks)).
+					Error("Storage is throttled. Scrapes and rule evaluations will be skipped.")
```
Comment (fabxc) [inline storage/local/storage.go]: As discussed yesterday, I'm not fully sure this is the place to log this "state" – but it's certainly the most convenient one implementation-wise right now.

If we have scrape intervals of greater than 1 minute this logging will be flapping. We don't really need to inform whether it stopped (since it's not an actual state anyway and the stopping message is unreliable as it depends on another call to `Throttle()`).
If the minutely logging goes away, that's enough to know in my opinion.

We also don't have to send on the channel on each call to `Throttle()` (might be a few hundred times/second) and consume each item.

To make this simpler, how about `s.throttle = make(chan struct{}, 1)`.

Then, in `Throttled()` so there is no potential blocking if the server is busy:
```go
select {
case s.throttled <- struct{}{}:
default:
    // Signal already pending.
}
```

In the consuming loop:
```go
timer := time.NewTimer(0)

for range timer.C {
    <-s.throttled
    timer.Reset(time.Minute)

    // Log the things.
}
```

File: storage/local/interface.go (hunk 2 above)
Comment (beorn7) [inline storage/local/interface.go]: `Throttle()` sounds more like a method to throttle the storage, not like a question. I was thinking about the inverted logic, but then it always sounds like the storage is incapable of ingesting more, which is not true. It just wants fewer samples so that it can recover from the overload.

How about `NeedsThrottling()`?

File: storage/storage.go (hunk 3 above)
Comment (beorn7) [inline storage/storage.go]: I'd express the contract as follows:

There are "best effort" appenders like the various remote storages, that will drop samples at will if they feel like it. They always return `false` on `NeedsThrottling()`. Then there are "guaranteed" appenders that will always append what they are asked for (and return an error otherwise once https://github.com/prometheus/prometheus/issues/1356 is fixed). They will use `NeedsThrottling()` to communicate that they need throttling. At the moment, we only have one "guaranteed" appender, the local storage. I cannot think of a reason to have more, so the discussion is somehow academic. But if you asked me, I'd say the whole reason to use a "guaranteed" appender is that we want all the samples there. So even if only a single one of them needs throttling, we will throttle.

> On Append() it should then only call Append() on those sub-appenders that are not busy.

This is impossible to implement. The Append() call is ignorant of where it is coming from (_und das ist auch gut so_). So for how long should it not propagate Append() calls to the overloaded appender?

In summary, I think the currently implemented logic is correct, and another logic cannot be implemented without changing the design more fundamentally.

File: storage/local/storage.go (hunk 4 above)
Comment (beorn7) [inline storage/local/storage.go]: I like the idea with `select` to skip if the channel has already been sent to. (But note that we wouldn't try to send on every call of Throttle, but only _if_ we are actually throttled, in which case we would not do a lot of harm by waiting as the caller will not start a scrape anyway.)

I thought long about not having the "stopping" log (following exactly your line of argument) and came to the conclusion it would be more confusing to the users to leave it out. Note that the message is _not_ repeated every minute. I invested some brain efforts to make it like the message for "rushed mode".

Note that the message will only be flapping if Throttle() is called less than once per minute. Which would hardly cause an overload of the storage in the first place. (Even if we are on a weak machine that can only ingest 10k samples/s, we needed scrapes that contain 600,000 metrics each to trigger that threshold with one call per minute.) Yes, in that pathological situation, you would get a flapping message, but you would probably have other problems, too.

**Comment by beorn7** [comment]
`select` tweak is applied, but I wait with the push until we know what to call the `Throttled` method.

**Comment by beorn7** [comment]
I'll go ahead an name it `NeedsThrottling()` and change the doc comment to make the contract clearer (as discussed above). Stay tuned...

File: storage/storage.go (hunk 3 above)
Comment (fabxc) [inline storage/storage.go]: We can throttle if a single one needs it – but that doesn't mean we have to throttle all the other ones.

It would be a bunch easier if `Append()` accepted a vector of samples.

Comment (beorn7) [inline storage/storage.go]: > We can throttle if a single one needs it – but that doesn't mean we have to throttle all the other ones.

This is not the way throttling works (at least with the basic idea implemented here). Throttling means that you don't even start a scrape if your storage is not ready to take the result.

> It would be a bunch easier if `Append()` accepted a vector of samples.

Which leads us back to the old discussion why we _removed_ a vector of samples from the Append signature.
* It requires to copy all samples (up to 500k/s) into a vector before handing it over to the storage, with no way of reusing it.
* It suggests atomicity that the current storage is not designed to deliver. (How wold you return the information that for three out of the 1000 samples in the vector, there was an out-of-order timestamp?)
* It does not help with the idea of _first_ finding out if the storage is ready to ingest and _then_ do the work of a potentially very large scrape.

Comment (fabxc) [inline storage/storage.go]: Well, if one storage can take it, I want to insert and consequentially do the work. I use throttle to determine whether the result of my work can be used. Which it can by one out of several storages. If none of them does, than I don't need to do the work.

> It requires to copy all samples (up to 500k/s) into a vector before handing it over to the storage, with no way of reusing it.

No. This is per scrape per target. Thus those vectors will be about ~1k samples on average and 10-20k in the extreme cases, which is not really an issue memory-wise. The vector of Sample pointers is perfectly reusable.

> It suggests atomicity that the current storage is not designed to deliver. (How wold you return the information that for three out of the 1000 samples in the vector, there was an out-of-order timestamp?)

A multi-error.
Guarantees can be documented.

> It does not help with the idea of first finding out if the storage is ready to ingest and then do the work of a potentially very large scrape.

Suppose N storages in a fanout. If one or more can take samples, do the scrape and call `Append(<all samples of the scrape>)`. Fanout calls `Append()` for all M storages that are not throttled.

**Comment by beorn7** [comment]
OK, updated. PHAL.

WRT multiple non-best-effort storages: I define the contract such that we want to append each and every sample to each non-best-effort storage. We could discuss if that's a reasonable assumption but since it will not happen in ages, it would be purely academic. Should the contract be such that appending to at least one non-best-effort storage is good enough, than your logic is sound. But coding that logic for the hypothetical case that we ever have more than one non-best-effort storage would be premature even if I could agree that your idea of the contract was the right one. Currently, it's perfectly clear that we don't want to do any work if our only one non-best-effort storage is not ready for ingestion.

WRT appending vectors: There are more reasons. I don't think this is the right place to discuss further. The logic of "first ask for throttling, then perform work" would still make sense. So even if we wanted to change back the storage interface, it's independent from this change and should be considered (carefully!) separately.

File: rules/manager.go [hunk 5]
```diff
@@ -69,6 +69,11 @@ var (
 		Help:       "The duration for all evaluations to execute.",
 		Objectives: map[float64]float64{0.01: 0.001, 0.05: 0.005, 0.5: 0.05, 0.90: 0.01, 0.99: 0.001},
 	})
+	iterationsSkipped = prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (fabxc) [inline rules/manager.go]: We have the total number of iteratoins via the summary above. It might still be clearer to have a seperate `evaluator_iterations_total`. (And maybe name this one `evaluator_skipped_iterations_total`).

**Comment by fabxc** [comment]
Agreed that we don't need to resolve this discussion for this PR.
We did agree on `Append()` returning an error though, right?

LGTM after that's added.

**Comment by beorn7** [comment]
Yes, we agreed to let `Append()` return an error, but that's another orthogonal thing to implement (and individual PRs should be as lean as possible – I got reminded of that with all the SC-internal (and SC-management-internal) recent discussions about mono-repos and how code reviews and branching should be dealt with). I have already filed #1356 to keep track of it.

File: rules/manager.go (hunk 5 above)
Comment (beorn7) [inline rules/manager.go]: I have reworded the help strings for more clarity (already anticipating grouping of rules).
I have also added the suggested total counter.

**Comment by beorn7** [comment]
Also removed that spurious "x" from the commit message now.

**Comment by fabxc** [comment]
Seperate PR it is.

:+1: :ship:
